src/v2/Components/FlashBanner/index.tsx

```diff
@@ -1,43 +1,103 @@
 import React from "react"
+import { Banner, Flex, Sans } from "@artsy/palette"
+import { QueryRenderer, graphql } from "react-relay"
+import { useSystemContext } from "v2/Artsy"
 import qs from "qs"
-import { Banner, Sans } from "@artsy/palette"
+import { EmailConfirmationCTA } from "./EmailConfirmationCTA"
+import { useDidMount } from "v2/Utils/Hooks/useDidMount"
 
-/** Flash messages triggered by a url `flash_message` query param
- * A message can be a string or a React Component that takes no props
- * (that is, use relay or context if necessary)
- */
-const messageByQueryParam: { [k: string]: string | React.FC } = {
-  confirmed: "Your email has been confirmed.",
-  already_confirmed: "You have already confirmed your email.",
-  invalid_token: "An error has occurred. Please contact support@artsy.net.",
-  blank_token: "An error has occurred. Please contact support@artsy.net.",
-  expired_token: "Link expired. Resend verification email.",
+interface FlashBannerProps {
+  contentCode?: string
+  me?: {
+    canRequestEmailConfirmation: boolean
+  }
+}
+
+type BannerContent = string | React.ComponentType
+
+function isComponent(c: BannerContent): c is React.ComponentType {
+  return typeof c === "function"
 }
 
 /**
- * The component responsible for selecting a flash message key from the `flash_message` url query param
+ * The component responsible for selecting a determining and displaying a flash message
+ *
  */
-export const FlashBanner: React.FunctionComponent<{
-  messageCode?: string
-}> = props => {
-  const [messageCode, setMessageCode] = React.useState(props.messageCode)
-
-  React.useEffect(() => {
-    if (!messageCode) {
-      const query = qs.parse(window.location.search.slice(1))
-      if (query["flash_message"]) setMessageCode(query["flash_message"])
-    }
-  }, [messageCode])
-
-  if (!messageCode) return null
-  const Message = messageByQueryParam[messageCode]
-  if (!Message) return null
+export const FlashBanner: React.FC<FlashBannerProps> = props => {
+  /**
+   * A map indexing keys (which may come from a contentCode prop, query string or
+   * logic internal to the component + its props) to banner content.
+   */
+  const contentMap: Record<string, BannerContent> = {
+    /* Comes from actual window.location.search query params */
+    confirmed: "Your email has been confirmed.",
+    already_confirmed: "You have already confirmed your email.",
+    invalid_token: "An error has occurred. Please contact support@artsy.net.",
+    blank_token: "An error has occurred. Please contact support@artsy.net.",
+    expired_token: "Link expired. Resend verification email.",
+
+    /* Comes from logic */
+    email_confirmation_cta: EmailConfirmationCTA,
```

Comment:
this section may feel sort of needlessly complicated - it's main purpose is to provide a single interface for 3 different ways that a message might be set, including one (the `contentCode` prop) that can be set to skip logic dependent on window or relay - that is, currently, for storybooks. The common interface (this object's keys) makes this easier - the keys are effectively constants.

Reply:
> this section may feel sort of needlessly complicated

Want to echo again my comment [here](https://github.com/artsy/force/pull/5667#discussion_r432779486), and in particular the importance of reducing complexity by eliminating logical branches in code as well as the type-system. 

This content map object has as low a level as complexity as it can _until_ one mixes in an entirely different data structure in the form of a react component. As mentioned in the link above, it's not strictly necessary that the email confirmation component reference should go goes here (its its own thing) and by removing we simplify the system in a number of ways:

- Reduces logic
- Eliminates the use of union types
- Eliminates helper functions for checking said union types
- Eliminates the mental overhead of needing in-line code comments describing an unexpected logical branch
- Eliminates the need for multi-purpose variable name casing (a simple string as `Uppercase`? -- or is a `Component`?)
- Eliminates the ternary in the render function for rending a string or a component. 

Not trying to be pedantic about it 😄 Just trying to emphasize that oftentimes the structures one creates to support simple-to-complex data structures in one go comes at cost, and to evaluate if these structures are really necessary. (Brings to mind Java and / or OOP with all of its generic base classes and so on, when really what we needs is simple function that does one thing and one thing only.)